src/client/features/admin/index.js

```diff
@@ -174,4 +175,195 @@ class Admin extends React.Component {
+  
+      this.state = {
+        query: _.assign({q: '', gt: 2, lt: 250, type: 'Pathway'}, query),
+        searchResults: [],
+        editLinks: [],
+        dataSources: []
+      };
+  
+      PathwayCommonsService.datasources()
+      .then(result => this.setState({
+        dataSources: result
+        })
+      );
+      this.updateEditLinks = this.updateEditLinks.bind(this);
+    }
+  
+    updateEditLinks(editURI) {
+      const editLinkObj = this.deconstructEditURI(editURI);
+      var editLinks = this.state.editLinks;
+      editLinks[editLinkObj.uri] = editLinkObj.editkey;
+      this.setState({
+        editLinks: editLinks
+      });
+    }
+  
+    deconstructEditURI(editURI) {
+      const id = '&editkey=';
+      const uri = editURI.substring(0, editURI.indexOf(id));
+      const editkey = editURI.substring(editURI.indexOf(id) + id.length, editURI.length);
+      return {uri: uri, editkey: editkey};
+    }
+
+    getSearchResult() {
+      const state = this.state;
+      const query = state.query;
+  
+      if (query.q !== '') {
+        PathwayCommonsService.querySearch(query)
+        .then(searchResults => {
+          this.setState({searchResults: searchResults}, () => {
+            CDC.initEditLinkSocket(this.updateEditLinks);
+            for (var i = 0; i < searchResults.length; i++) {
+              CDC.requestEditLink(searchResults[i].uri, 'latest');
+            }
+          });
+        });
+      }
+    }
+  
+    componentDidMount() {
+      this.getSearchResult();
+    }
+  
+    onSearchValueChange(e) {
+      // if the user presses enter, submit the query
+      if (e.which && e.which ===  13) {
+        this.submitSearchQuery(e);
+      } else {
+        const newQueryState = _.assign({}, this.state.query);
+        newQueryState.q = e.target.value;
+        this.setState({query: newQueryState});
+      }
+    }
+  
+    setQueryType(e, type) {
+      const newQueryState = _.assign({}, this.state.query);
+      newQueryState.type = type;
+      this.setState({query: newQueryState}, function () { this.submitSearchQuery(); });
+    }
+  
+    submitSearchQuery() {
+      const props = this.props;
+      const state = this.state;
+  
+      const query = state.query;
+      const uriRegex = /^(?:http(s)?:\/\/)?[\w.-]+(?:\.[\w\.-]+)+[\w\-\._~:/?#[\]@!\$&'\(\)\*\+,;=.]+$/;
+  
+      if (query.q.match(uriRegex)) {
+        CDC.initEditLinkSocket((editLink) => {
+          const editLinkObj = this.deconstructEditURI(editLink);
+          props.history.push({
+            pathname: '/view',
+            search: queryString.stringify({uri: editLinkObj.uri, editkey: editLinkObj.editkey}),
+            state: {}
+          });
+        });
+        CDC.requestEditLink(query.q, 'latest');
+      } else {
+        props.history.push({
+          pathname: '/admin',
+          search: queryString.stringify(query),
+          state: {}
+        });
+  
+        this.getSearchResult();
+      }
```

Comment:
Just to clarify.  This admin thing is a copy of some of the search stuff I have been working on with some extra functionality like providing admin edit links I think(correct me if I am wrong Jonah). 

In the latest version of the search code, I disable most ui elements that submit search queries, but I don't disable the search bar, so that is one point where concurrent queries can be an issue.

I do have a loading flag in this.state that I use to display a loading indicator.  Do you think it makes sense to do something like:

```js

getSearchResult() {
   const state = this.state;
   if (state.query.q !== '' && !state.loading) {
      // perform search
      ...
      this.setState({loading: false});
      ...
   }

}
```

I think this would allow only one search request to be performed at a time, regardless of ui enabled/disabled state.

Reply:
https://github.com/PathwayCommons/app-ui/blob/master/src/client/features/search/index.js#L42

Reply:
Much of the admin app was copied directly from the search app, yes. I can always re-copy the search app and re-add the admin edit link functionality at any time.